File: auth-service/src/main/java/com/rockrager/authentication/service/AuthService.java
package com.rockrager.authentication.service;

import com.rockrager.authentication.dto.request.LoginRequest;
import com.rockrager.authentication.dto.request.RegisterRequest;
import com.rockrager.authentication.dto.response.AuthResponse;
import com.rockrager.authentication.dto.response.LoginInitiateResponse;
import com.rockrager.authentication.entity.*;
import com.rockrager.authentication.repository.EmailVerificationTokenRepository;
import com.rockrager.authentication.repository.PasswordResetTokenRepository;
import com.rockrager.authentication.repository.RefreshTokenRepository;
import com.rockrager.authentication.repository.UserRepository;
import com.rockrager.authentication.security.jwt.JwtService;
import com.rockrager.authentication.repository.OtpCodeRepository;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final RefreshTokenRepository refreshTokenRepository;
    private final EmailVerificationTokenRepository emailVerificationTokenRepository;
    private final PasswordResetTokenRepository passwordResetTokenRepository;
    private final OtpCodeRepository otpCodeRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final EmailService emailService;
    private final OtpService otpService;
    private final DeviceInfoService deviceInfoService;

    @Transactional
    public AuthResponse register(RegisterRequest request) {
        log.info("Registering user with email: {}", request.getEmail());

        if (userRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("Email already registered");
        }

        User user = User.builder()
                .firstName(request.getFirstName())
                .lastName(request.getLastName())
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .emailVerified(false)
                .role("USER")
                .loginCount(0)
                .otpEnabled(true)
                .authProvider(AuthProvider.LOCAL)
                .createdAt(LocalDateTime.now())
                .updatedAt(LocalDateTime.now())
                .build();

        User savedUser = userRepository.save(user);
        log.info("User saved with ID: {}", savedUser.getId());

        String accessToken = jwtService.generateAccessToken(savedUser.getEmail());
        String refreshToken = jwtService.generateRefreshToken(savedUser.getEmail());

        refreshTokenRepository.deleteByUser(savedUser);

        RefreshToken refreshTokenEntity = RefreshToken.builder()
                .user(savedUser)
                .token(refreshToken)
                .expiresAt(LocalDateTime.now().plusDays(7))
                .revoked(false)
                .build();

        refreshTokenRepository.save(refreshTokenEntity);

        String verificationToken = UUID.randomUUID().toString();

        EmailVerificationToken emailVerificationTokenEntity = EmailVerificationToken.builder()
                .token(verificationToken)
                .user(savedUser)
                .expiresAt(LocalDateTime.now().plusHours(24))
                .build();

        emailVerificationTokenRepository.save(emailVerificationTokenEntity);

        try {
            emailService.sendVerificationEmail(
                    savedUser.getEmail(),
                    savedUser.getFirstName(),
                    verificationToken
            );
            log.info("Verification email sent to: {}", savedUser.getEmail());
        } catch (Exception e) {
            log.error("Failed to send verification email to: {}", savedUser.getEmail(), e);
        }

        return AuthResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .message("User registered successfully. Please check your email for verification link.")
                .build();
    }

    @Transactional
    public AuthResponse login(LoginRequest request) {

        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new RuntimeException("Invalid email or password"));

        if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            throw new RuntimeException("Invalid email or password");
        }

        if (!user.isEmailVerified()) {
            throw new RuntimeException("Please verify your email first. Check your inbox for verification link.");
        }

        boolean hasLoggedInBefore = user.getLoginCount() > 0;

        log.info("User login attempt for: {}, Login count: {}, hasLoggedInBefore: {}",
                user.getEmail(), user.getLoginCount(), hasLoggedInBefore);

        if (hasLoggedInBefore) {
            log.info("Existing user - returning requiresOtp: true for: {}", user.getEmail());

            String maskedEmail = maskEmail(user.getEmail());

            return AuthResponse.builder()
                    .requiresOtp(true)
                    .message("Please verify your identity")
                    .email(maskedEmail)
                    .build();
        } else {
            log.info("First time login for user: {}", user.getEmail());

            String accessToken = jwtService.generateAccessToken(user.getEmail());
            String refreshToken = jwtService.generateRefreshToken(user.getEmail());

            refreshTokenRepository.deleteByUser(user);

            RefreshToken refreshTokenEntity = RefreshToken.builder()
                    .user(user)
                    .token(refreshToken)
                    .expiresAt(LocalDateTime.now().plusDays(7))
                    .revoked(false)
                    .build();

            refreshTokenRepository.save(refreshTokenEntity);

            user.setLoginCount(user.getLoginCount() + 1);
            user.setLastLoginAt(LocalDateTime.now());
            user.setLastLoginDevice(request.getDeviceInfo());
            user.setLastLoginIp(request.getIpAddress());
            userRepository.save(user);

            return AuthResponse.builder()
                    .accessToken(accessToken)
                    .refreshToken(refreshToken)
                    .message("Login successful")
                    .build();
        }
    }

    @Transactional
    public AuthResponse refreshToken(String refreshToken) {

        RefreshToken storedToken = refreshTokenRepository
                .findByToken(refreshToken)
                .orElseThrow(() -> new RuntimeException("Invalid refresh token"));

        if (storedToken.isRevoked()) {
            throw new RuntimeException("Refresh token revoked");
        }

        if (storedToken.getExpiresAt().isBefore(LocalDateTime.now())) {
            refreshTokenRepository.delete(storedToken);
            throw new RuntimeException("Refresh token expired");
        }

        User user = storedToken.getUser();

        String newAccessToken = jwtService.generateAccessToken(user.getEmail());

        return AuthResponse.builder()
                .accessToken(newAccessToken)
                .refreshToken(refreshToken)
                .message("Access token refreshed")
                .build();
    }

    @Transactional
    public String logout(String refreshToken) {

        RefreshToken token = refreshTokenRepository
                .findByToken(refreshToken)
                .orElseThrow(() -> new RuntimeException("Invalid refresh token"));

        token.setRevoked(true);
        refreshTokenRepository.save(token);

        return "Logout successful";
    }

    @Transactional
    public String verifyEmail(String token) {
        EmailVerificationToken verificationToken = emailVerificationTokenRepository
                .findByToken(token)
                .orElseThrow(() -> new RuntimeException("Invalid verification token"));

        if (verificationToken.getExpiresAt().isBefore(LocalDateTime.now())) {
            emailVerificationTokenRepository.delete(verificationToken);
            throw new RuntimeException("Verification token expired. Please register again.");
        }

        User user = verificationToken.getUser();
        user.setEmailVerified(true);
        userRepository.save(user);

        emailVerificationTokenRepository.delete(verificationToken);

        log.info("Email verified successfully for user: {} ({})", user.getEmail(), user.getFirstName());

        sendWelcomeEmailWithRetry(user.getEmail(), user.getFirstName(), 3);

        return "Email verified successfully. You can now login.";
    }

    private void sendWelcomeEmailWithRetry(String email, String firstName, int maxRetries) {
        for (int attempt = 1; attempt <= maxRetries; attempt++) {
            try {
                emailService.sendWelcomeEmail(email, firstName);
                log.info("Welcome email sent successfully to: {} on attempt {}", email, attempt);
                return;
            } catch (Exception e) {
                log.warn("Failed to send welcome email to: {} on attempt {}/{}", email, attempt, maxRetries, e);
                if (attempt == maxRetries) {
                    log.error("Failed to send welcome email to: {} after {} attempts", email, maxRetries);
                    storeFailedEmailNotification(email, firstName, "WELCOME");
                }
                try {
                    Thread.sleep(1000 * attempt);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
        }
    }

    private void storeFailedEmailNotification(String email, String firstName, String emailType) {
        log.info("Storing failed email notification for: {} of type: {}", email, emailType);
    }

    @Transactional
    public String forgotPassword(String email) {

        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found with email: " + email));

        passwordResetTokenRepository.findByToken(email).ifPresent(existingToken ->
                passwordResetTokenRepository.delete(existingToken)
        );

        String token = UUID.randomUUID().toString();

        PasswordResetToken resetToken = PasswordResetToken.builder()
                .token(token)
                .user(user)
                .expiresAt(LocalDateTime.now().plusHours(1))
                .build();

        passwordResetTokenRepository.save(resetToken);

        try {
            emailService.sendPasswordResetEmail(
                    user.getEmail(),
                    user.getFirstName(),
                    token
            );
            log.info("Password reset email sent to: {} ({})", user.getEmail(), user.getFirstName());
        } catch (Exception e) {
            log.error("Failed to send password reset email to: {}", user.getEmail(), e);
            throw new RuntimeException("Failed to send password reset email. Please try again.");
        }

        return "Password reset instructions sent to your email. Please check your inbox.";
    }

    @Transactional
    public String resetPassword(String token, String newPassword) {

        PasswordResetToken resetToken = passwordResetTokenRepository
                .findByToken(token)
                .orElseThrow(() -> new RuntimeException("Invalid or expired reset token"));

        if (resetToken.getExpiresAt().isBefore(LocalDateTime.now())) {
            passwordResetTokenRepository.delete(resetToken);
            throw new RuntimeException("Reset token has expired. Please request a new one.");
        }

        User user = resetToken.getUser();

        user.setPassword(passwordEncoder.encode(newPassword));
        userRepository.save(user);

        passwordResetTokenRepository.delete(resetToken);

        refreshTokenRepository.deleteByUser(user);

        log.info("Password reset successful for user: {}", user.getEmail());

        return "Password reset successful. Please login with your new password.";
    }

    @Transactional
    public LoginInitiateResponse initiateLogin(LoginRequest request) {
        log.info("INITIATE LOGIN - START");

        String originalEmail = request.getEmail();
        log.info("Original email for lookup: {}", originalEmail);

        log.info("Request password length: {}", request.getPassword() != null ? request.getPassword().length() : 0);
        log.info("Request device info: {}", request.getDeviceInfo());
        log.info("Request IP address: {}", request.getIpAddress());

        log.info("Searching for user with email: {}", originalEmail);
        User user = userRepository.findByEmail(originalEmail)
                .orElseThrow(() -> {
                    log.error("User not found with email: {}", originalEmail);
                    return new RuntimeException("Invalid email or password");
                });

        log.info("User found - ID: {}, Email: {}, Login Count: {}, Auth Provider: {}",
                user.getId(), user.getEmail(), user.getLoginCount(), user.getAuthProvider());

        if (user.getAuthProvider() == AuthProvider.GOOGLE && (user.getPassword() == null || user.getPassword().isEmpty())) {
            log.warn("User {} is Google-only account with no password", user.getEmail());
            throw new RuntimeException("This account uses Google login. Please sign in with Google.");
        }

        if (user.getGoogleId() != null && user.getAuthProvider() == AuthProvider.LOCAL) {
            log.info("User with linked Google account logging in with password: {}", request.getEmail());
        }

        log.info("Validating password...");
        boolean passwordMatches = passwordEncoder.matches(request.getPassword(), user.getPassword());
        log.info("Password matches: {}", passwordMatches);

        if (!passwordMatches) {
            log.error("Password mismatch for user: {}", originalEmail);
            throw new RuntimeException("Invalid email or password");
        }
        log.info("Password validated successfully");

        log.info("Checking email verification status...");
        if (!user.isEmailVerified()) {
            log.warn("User {} has not verified email", user.getEmail());
            throw new RuntimeException("Please verify your email first. Check your inbox for verification link.");
        }
        log.info("Email is verified");

        otpCodeRepository.deleteByUserAndUsedFalse(user);
        log.info("Cleared any existing unused OTPs for user: {}", user.getEmail());

        String sessionId = UUID.randomUUID().toString();
        log.info("Generated session ID: {}", sessionId);

        log.info("Generating and sending OTP...");
        String otpCode = otpService.generateAndSendOtp(
                user,
                sessionId,
                request.getDeviceInfo(),
                request.getIpAddress()
        );
        log.info("OTP generated and sent - Code: {}", otpCode);

        log.info("OTP sent to user: {} for session: {}", user.getEmail(), sessionId);

        String maskedEmail = maskEmail(user.getEmail());
        log.info("Masked email for response: {}", maskedEmail);

        log.info("INITIATE LOGIN - SUCCESS");

        return LoginInitiateResponse.builder()
                .sessionId(sessionId)
                .otpRequired(true)
                .message("OTP sent to your email address")
                .otpSentTo(maskedEmail)
                .expiresIn((long) otpService.getOtpExpirySeconds())
                .build();
    }

    @Transactional
    public AuthResponse verifyOtpAndLogin(com.rockrager.authentication.dto.request.OtpVerificationRequest request) {
        log.info("Verifying OTP for session: {}", request.getSessionId());

        boolean isValid = otpService.validateOtp(request.getSessionId(), request.getOtpCode());

        if (!isValid) {
            throw new RuntimeException("Invalid or expired OTP. Please try again.");
        }

        com.rockrager.authentication.entity.OtpCode otpRecord = otpService.getOtpRecord(request.getSessionId())
                .orElseThrow(() -> new RuntimeException("Session not found"));

        User user = otpRecord.getUser();

        if (user.getAuthProvider() == AuthProvider.GOOGLE && (user.getPassword() == null || user.getPassword().isEmpty())) {
            throw new RuntimeException("This account uses Google login. Please sign in with Google.");
        }

        user.setLastLoginAt(LocalDateTime.now());
        user.setLastLoginIp(otpRecord.getIpAddress());
        user.setLastLoginDevice(otpRecord.getDeviceInfo());

        user.setLoginCount(user.getLoginCount() + 1);

        try {
            String location = deviceInfoService.getLocationFromIp(otpRecord.getIpAddress());
            user.setLastLoginLocation(location);
        } catch (Exception e) {
            log.warn("Could not get location for IP: {}", otpRecord.getIpAddress());
            user.setLastLoginLocation("Unknown");
        }

        userRepository.save(user);

        String accessToken = jwtService.generateAccessToken(user.getEmail());
        String refreshToken = jwtService.generateRefreshToken(user.getEmail());

        refreshTokenRepository.deleteByUser(user);
        RefreshToken refreshTokenEntity = RefreshToken.builder()
                .user(user)
                .token(refreshToken)
                .expiresAt(LocalDateTime.now().plusDays(7))
                .revoked(false)
                .build();
        refreshTokenRepository.save(refreshTokenEntity);

        try {
            sendLoginNotificationEmail(user, otpRecord);
        } catch (Exception e) {
            log.error("Failed to send login notification email to: {}", user.getEmail(), e);
        }

        otpService.cleanupExpiredOtps(user);

        log.info("User logged in successfully: {} from IP: {}", user.getEmail(), otpRecord.getIpAddress());

        return AuthResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .message("Login successful")
                .build();
    }

    private void sendLoginNotificationEmail(User user, com.rockrager.authentication.entity.OtpCode otpRecord) {
        String subject = "New Login Detected - RockRager Authentication";

        String deviceInfo = otpRecord.getDeviceInfo() != null ? otpRecord.getDeviceInfo() : "Unknown Device";
        String ipAddress = otpRecord.getIpAddress() != null ? otpRecord.getIpAddress() : "Unknown IP";
        String location = user.getLastLoginLocation() != null ? user.getLastLoginLocation() : "Unknown Location";
        String loginTime = LocalDateTime.now().toString();

        String body = String.format("""
        Hello %s %s,
        
        We detected a new login to your account.
        
        Login Details:
        • Time: %s
        • IP Address: %s
        • Location: %s
        • Device: %s
        
        If this was you, you can ignore this email.
        
        If this wasn't you, please reset your password immediately and contact support.
        
        Best regards,
        RockRager Team
        """,
                user.getFirstName(),
                user.getLastName(),
                loginTime,
                ipAddress,
                location,
                deviceInfo
        );

        try {
            emailService.sendLoginNotificationEmail(user.getEmail(), user.getFirstName(), subject, body);
        } catch (Exception e) {
            log.error("Failed to send login notification", e);
            throw e;
        }
    }

    private String maskEmail(String email) {
        if (email == null || !email.contains("@")) {
            return email;
        }
        String[] parts = email.split("@");
        String localPart = parts[0];
        String domain = parts[1];

        if (localPart.length() <= 2) {
            return "*" + "@" + domain;
        }

        String maskedLocal = localPart.substring(0, 2) + "***" + localPart.substring(localPart.length() - 1);
        return maskedLocal + "@" + domain;
    }
}

File: client/src/features/auth/services/authService.ts
import apiClient from '../../../services/api/apiClient';
import { endpoints } from '../../../services/api/endpoints';
import { tokenManager } from '../utils/tokenManager';
import { STORAGE_KEYS } from '../../../constants/app.constants';
import {
  RegisterData,
  RegisterResponse,
  LoginCredentials,
  LoginResponse,
  OtpVerifyRequest,
  OtpVerifyResponse,
  ForgotPasswordRequest,
  ResetPasswordRequest,
  User,
} from '../types/auth.types';

class AuthService {
  // User Registration
  async register(data: RegisterData): Promise<RegisterResponse> {
    console.log('AuthService.register called with:', data);
    
    try {
      const response = await apiClient.post<RegisterResponse>(
        endpoints.auth.register,
        data
      );
      
      console.log('Registration response:', response.data);
      
      // Store access token if provided
      if (response.data.accessToken) {
        tokenManager.setAccessToken(response.data.accessToken);
        console.log('Access token stored');
      }
      
      return response.data;
    } catch (error: any) {
      console.error('Registration error:', error.response?.data || error.message);
      throw error;
    }
  }
  
  // Email Verification (POST method)
  async verifyEmail(token: string): Promise<string> {
    console.log('AuthService.verifyEmail called with token:', token);
    
    try {
      const response = await apiClient.post<string>(
        endpoints.auth.verifyEmail,
        { token }
      );
      console.log('Verify email response:', response.data);
      return response.data;
    } catch (error: any) {
      console.error('Verify email error:', error.response?.data || error.message);
      throw error;
    }
  }
  
  // Login
  async login(credentials: LoginCredentials): Promise<LoginResponse> {
    console.log('AuthService.login called with:', credentials.email);
    
    try {
      const response = await apiClient.post<LoginResponse>(
        endpoints.auth.login,
        credentials
      );
      
      console.log('Login response:', response.data);
      
      // If direct login (no OTP required) - store token
      if (response.data.accessToken) {
        tokenManager.setAccessToken(response.data.accessToken);
        console.log('Access token stored');
      }
      
      // If OTP required - store session data
      if (response.data.requiresOtp && response.data.sessionId) {
        tokenManager.setSessionId(response.data.sessionId);
        if (response.data.expiresIn) {
          tokenManager.setOtpExpiry(response.data.expiresIn);
        }
        console.log('OTP session stored');
      }
      
      return response.data;
    } catch (error: any) {
      console.error('Login error:', error.response?.data || error.message);
      throw error;
    }
  }
  
  // Verify OTP
  async verifyOtp(otpCode: string): Promise<OtpVerifyResponse> {
    const sessionId = tokenManager.getSessionId();
    console.log('AuthService.verifyOtp called with sessionId:', sessionId, 'otpCode:', otpCode);
    
    if (!sessionId) {
      console.error('No session ID found');
      throw new Error('Session expired. Please login again.');
    }
    
    const request: OtpVerifyRequest = { sessionId, otpCode };
    
    try {
      const response = await apiClient.post<OtpVerifyResponse>(
        endpoints.auth.verifyOtp,
        request
      );
      
      console.log('Verify OTP response:', response.data);
      
      // Store access token on successful OTP verification
      if (response.data.accessToken) {
        tokenManager.setAccessToken(response.data.accessToken);
        // Clear OTP session data
        tokenManager.removeSessionId();
        tokenManager.removeOtpExpiry();
        console.log('Access token stored, OTP session cleared');
      }
      
      return response.data;
    } catch (error: any) {
      console.error('Verify OTP error:', error.response?.data || error.message);
      throw error;
    }
  }
  
  // Forgot Password
  async forgotPassword(email: string): Promise<string> {
    console.log('AuthService.forgotPassword called with:', email);
    
    try {
      const request: ForgotPasswordRequest = { email };
      const response = await apiClient.post<string>(
        endpoints.auth.forgotPassword,
        request
      );
      console.log('Forgot password response:', response.data);
      return response.data;
    } catch (error: any) {
      console.error('Forgot password error:', error.response?.data || error.message);
      throw error;
    }
  }
  
  // Reset Password
  async resetPassword(token: string, newPassword: string): Promise<string> {
    console.log('AuthService.resetPassword called with token:', token);
    
    try {
      const request: ResetPasswordRequest = { token, newPassword };
      const response = await apiClient.post<string>(
        endpoints.auth.resetPassword,
        request
      );
      console.log('Reset password response:', response.data);
      return response.data;
    } catch (error: any) {
      console.error('Reset password error:', error.response?.data || error.message);
      throw error;
    }
  }
  
  // Refresh Token
  async refreshToken(): Promise<string | null> {
    console.log('AuthService.refreshToken called');
    
    try {
      const response = await apiClient.post<{ accessToken: string }>(
        endpoints.auth.refreshToken,
        {}
      );
      
      if (response.data.accessToken) {
        tokenManager.setAccessToken(response.data.accessToken);
        console.log('Token refreshed successfully');
        return response.data.accessToken;
      }
      return null;
    } catch (error) {
      console.error('Refresh token error:', error);
      return null;
    }
  }
  
  // Logout
  async logout(): Promise<string> {
    console.log('AuthService.logout called');
    
    try {
      const response = await apiClient.post<string>(endpoints.auth.logout, {});
      console.log('Logout response:', response.data);
      return response.data;
    } finally {
      // Always clear local tokens
      tokenManager.clearAll();
      localStorage.removeItem(STORAGE_KEYS.USER);
      console.log('All tokens cleared');
    }
  }
  
  // Get Current User
  async getCurrentUser(): Promise<User> {
    console.log('AuthService.getCurrentUser called');
    
    try {
      const response = await apiClient.get<{ user: User }>(
        endpoints.auth.getMe
      );
      
      console.log('Get current user response:', response.data);
      
      // Store user in localStorage for persistence
      localStorage.setItem(STORAGE_KEYS.USER, JSON.stringify(response.data.user));
      
      return response.data.user;
    } catch (error: any) {
      console.error('Get current user error:', error.response?.data || error.message);
      throw error;
    }
  }
  
  // Get stored user (from localStorage)
  getStoredUser(): User | null {
    const userStr = localStorage.getItem(STORAGE_KEYS.USER);
    if (userStr) {
      try {
        const user = JSON.parse(userStr);
        console.log('Stored user retrieved:', user);
        return user;
      } catch {
        return null;
      }
    }
    return null;
  }
  
  // Check if user is authenticated
  isAuthenticated(): boolean {
    const isAuth = tokenManager.isAuthenticated();
    console.log('isAuthenticated check:', isAuth);
    return isAuth;
  }
}

export const authService = new AuthService();
export default authService;